animations-dbus/animations-dbus-server.c

```diff
@@ -0,0 +1,624 @@
+  g_clear_pointer (&priv->animatable_surfaces, g_ptr_array_unref);
+
+  G_OBJECT_CLASS (animations_dbus_server_parent_class)->dispose (object);
+}
+
+static void
+animations_dbus_server_finalize (GObject *object)
+{
+  AnimationsDbusServer *server = ANIMATIONS_DBUS_SERVER (object);
+  AnimationsDbusServerPrivate *priv = animations_dbus_server_get_instance_private (server);
+
+  g_clear_pointer (&priv->client_name_watches, g_hash_table_unref);
+
+  if (priv->name_id != 0)
+    {
+      g_bus_unown_name (priv->name_id);
+      priv->name_id = 0;
+    }
+
+  G_OBJECT_CLASS (animations_dbus_server_parent_class)->finalize (object);
+}
+
+static void
+animations_dbus_server_init (AnimationsDbusServer *server)
+{
+  AnimationsDbusServerPrivate *priv = animations_dbus_server_get_instance_private (server);
+
+  priv->animatable_surfaces = g_ptr_array_new_with_free_func (g_object_unref);
+  priv->animation_managers = g_hash_table_new_full (g_str_hash,
+                                                    g_str_equal,
+                                                    g_free,
+                                                    g_object_unref);
+  priv->animation_manager_ids = g_hash_table_new_full (g_direct_hash,
+                                                       g_direct_equal,
+                                                       NULL,
+                                                       g_free);
+  priv->client_name_watches = g_hash_table_new_full (g_str_hash,
+                                                     g_str_equal,
+                                                     g_free,
+                                                     NULL);
+}
+
+static void
+animations_dbus_server_class_init (AnimationsDbusServerClass *klass)
+{
+  GObjectClass *object_class = G_OBJECT_CLASS (klass);
+
+  object_class->get_property = animations_dbus_server_get_property;
+  object_class->set_property = animations_dbus_server_set_property;
+  object_class->dispose = animations_dbus_server_dispose;
+  object_class->finalize = animations_dbus_server_finalize;
+
+  animations_dbus_server_props[PROP_EFFECT_FACTORY] =
+    g_param_spec_object ("effect-factory",
+                         "An AnimationsDbusServerEffectFactory",
+                         "The AnimationsDbusServerEffectFactory that creates the animations",
+                         ANIMATIONS_DBUS_TYPE_SERVER_EFFECT_FACTORY,
+                         G_PARAM_READWRITE | G_PARAM_CONSTRUCT_ONLY);
+
+  g_object_class_install_properties (object_class,
+                                     NPROPS,
+                                     animations_dbus_server_props);
+}
+
+void
+animations_dbus_server_new_async (AnimationsDbusServerEffectFactory *factory,
+                                  GCancellable                      *cancellable,
+                                  GAsyncReadyCallback                callback,
+                                  gpointer                           user_data)
```

Comment:
I added a `_finish` method shortly after this review was done (I'll document the other changes too).

I suppose that we could allow clients to pass their own connection ... my thought here was that the server object could exist independently of any other connection. Is it considered bad practice to have multiple connections from the same app?

Reply:
Generally you’d want to use a single connection to the bus from a single app. The only cases where I’ve ever done differently are:
 * When I’m connecting to the session bus *and* the system bus
 * Where I want to authenticate differently to the same D-Bus service (since authentication is typically based on the client connection’s unique name)

Reply:
Makes sense. I realized that having it accept a connection would be useful for the tests too, so lets do that.

Reply:
> I realized that having it accept a connection would be useful for the tests too,

Yes, very much so. You don’t want to tie the tests down to a particular bus instance (definitely not the system bus, ideally not the user’s real session bus either).